### PR TITLE
Enforce the amount of services and IRQs per partition

### PR DESCRIPTION
**Pull request by orenc17**
This change is [Reviewable](https://reviewable.io/reviews/kfnta/mbed-os/46)

### CODE REVIEW

File: tools/spm/generate_partition_code.py [hunk 1]
```diff
@@ -257,26 +258,32 @@ def __init__(
...
-            'Manifest {} - {} does not contain at least 1 IRQ or ROT_SRV'.format(
-                self.name, self.file
+        total_signals = len(self.rot_services) + len(self.irqs)
+        assert total_signals <= 32 - self.RESERVED_SIGNALS, \
+            'Manifest {} - {} exceeds limit of RoT services and IRQs allowed ' \
+            '({}).'.format(
+                self.name, self.file, 32 - self.RESERVED_SIGNALS
             )
 
+        assert total_signals > 0, 'Manifest {} - {} does not contain at least '\
```
Comment (mikisch81) [inline tools/spm/generate_partition_code.py]: "must contain" sounds better

File: tools/test/spm/test_data.py [hunk 2]
```diff
@@ -493,3 +492,225 @@
     ]
 }
 '''
+
+exceeding_services = [
```
Comment (mikisch81) [inline tools/test/spm/test_data.py]: I guess these 200+ lines can be squeezed to an expression due to the incremental nature of the dictionaries values,,

File: tools/spm/generate_partition_code.py (hunk 1 above)
Comment (mikisch81) [inline tools/spm/generate_partition_code.py]: Please check [https://github.com/ARMmbed/PSA-IPC-doc/issues/138](https://github.com/ARMmbed/PSA-IPC-doc/issues/138), they add a validation in the schema for it:
```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "description": "schema for a partition description.",
  "type": "object",
  "required": ["name", "type", "priority", "id", "entry_point", "stack_size", "heap_size", "source_files"],
  "anyOf": [
    {"required" : ["services"]},
    {"required" : ["irqs"]}
  ],
...
}
```

The schema validation will fail if there isn't any rot_service or irq in the json file.

File: tools/spm/generate_partition_code.py [hunk 3]
```diff
@@ -161,6 +161,7 @@ class Manifest(object):
         'HIGH': 'osPriorityHigh'
     }
     PARTITION_TYPES = ['APPLICATION-ROT', 'PSA-ROT']
+    RESERVED_SIGNALS = 5
```
Comment (mikisch81) [inline tools/spm/generate_partition_code.py]: Can you remind me why 5?
As I remember:
1-2: reserved
3: doorbell
32: not used

**Comment by orenc17** [comment]
Review status: 0 of 4 files reviewed, 3 unresolved discussions (waiting on @mikisch81 and @alzix)

---

*[tools/spm/generate_partition_code.py, line 164 at r1](https://reviewable.io/reviews/kfnta/mbed-os/46#-LG0CEUASp64Is-afhSu-r1-164:-LG17SwoA6EsgjVWQr-q:b-896fix) ([raw file](https://github.com/kfnta/mbed-os/blob/fed3dd62333f9b26afa7a4e9c9468fbee56074ff/tools/spm/generate_partition_code.py#L164)):*
<details><summary><i>Previously, mikisch81 (Michael Schwarcz) wrote…</i></summary><blockquote>

Can you remind me why 5?
As I remember:
1-2: reserved
3: doorbell
32: not used
</blockquote></details>

Done.

---

*[tools/spm/generate_partition_code.py, line 268 at r1](https://reviewable.io/reviews/kfnta/mbed-os/46#-LG0CEUASp64Is-afhSu-r1-268:-LG17UDJ0N4CplpYY9I-:b-896fix) ([raw file](https://github.com/kfnta/mbed-os/blob/fed3dd62333f9b26afa7a4e9c9468fbee56074ff/tools/spm/generate_partition_code.py#L268)):*
<details><summary><i>Previously, mikisch81 (Michael Schwarcz) wrote…</i></summary><blockquote>

Please check [https://github.com/ARMmbed/PSA-IPC-doc/issues/138](https://github.com/ARMmbed/PSA-IPC-doc/issues/138), they add a validation in the schema for it:
```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "description": "schema for a partition description.",
  "type": "object",
  "required": ["name", "type", "priority", "id", "entry_point", "stack_size", "heap_size", "source_files"],
  "anyOf": [
    {"required" : ["services"]},
    {"required" : ["irqs"]}
  ],
...
}
```

The schema validation will fail if there isn't any rot_service or irq in the json file.
</blockquote></details>

Done.

---

*[tools/test/spm/test_data.py, line 496 at r1](https://reviewable.io/reviews/kfnta/mbed-os/46#-LG0CEUASp64Is-afhSv-r1-496:-LG17VRQEMUpW3ntfJvi:b-896fix) ([raw file](https://github.com/kfnta/mbed-os/blob/fed3dd62333f9b26afa7a4e9c9468fbee56074ff/tools/test/spm/test_data.py#L496)):*
<details><summary><i>Previously, mikisch81 (Michael Schwarcz) wrote…</i></summary><blockquote>

I guess these 200+ lines can be squeezed to an expression due to the incremental nature of the dictionaries values,,
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kfnta/mbed-os/46)*

**Comment by alzix** [comment]
Reviewed 1 of 3 files at r1, 2 of 3 files at r3, 1 of 1 files at r4.
Dismissed @mikisch81 from 3 discussions.
Review status: :shipit: complete! all files reviewed, all discussions resolved

---



*Comments from [Reviewable](https://reviewable.io/reviews/kfnta/mbed-os/46)*

**Comment by alzix** [comment]
:lgtm:

---







*Comments from [Reviewable](https://reviewable.io/reviews/kfnta/mbed-os/46#-:-LG19tgR8wjoug-KXed2:bnfp4nl)*

**Pull request merged**